new_study_assistant.py

```diff
@@ -0,0 +1,624 @@
+
+        #with doc.create(Subsection('Number of unique attributes')):
+            with doc.create(Figure(position='h!')) as n_unique_attributes:
+                n_unique_attributes.add_image(file_directory + 'n_unique_attribute_distribution.png', width=NoEscape(r'0.99\textwidth'))
+                n_unique_attributes.add_caption('Comparison between the number of unique attributes in the new study and the number of unique attributes in each existing study on cBioPortal.  The dashed black line indicates the number of unqiue attributes in the new study, while the histogram shows the data for existing cBioPortal studies.')
+
+        #with doc.create(Subsection('Number of common attributes')):
+            with doc.create(Figure(position='h!')) as n_common_attributes:
+                n_common_attributes.add_image(file_directory + 'n_common_attribute_distribution.png', width=NoEscape(r'0.99\textwidth'))
+                n_common_attributes.add_caption('Comparison between the number of common attributes in the new study and the number of common attributes in each existing study on cBioPortal.  The dashed black line indicates the number of common attributes in the new study, while the histogram shows the data for existing cBioPortal studies.')
+
+    doc.generate_pdf('report', clean_tex=True)
+
+def combine_matches(exact, name, value, no_matches):   
+    all_matches = (pd.concat([exact, value, name, no_matches])
+                      .drop_duplicates()
+                      .sort_values('New study attribute',axis=0)
+                      .reset_index(drop=True))
+    
+    for index, row in all_matches.iterrows():
+        append_string = ''
+        if (sum(exact['Possible matches'] == row['Possible matches']) > 0) or (sum(name['Possible matches'] == row['Possible matches']) > 0):
+            append_string += '^'
+        if sum(value['Possible matches'] == row['Possible matches']) > 0:
+            append_string += '*'
+        else:
+            append_string += ''
+
+        if row['Possible matches'] != 'No matches found':            
+            all_matches['Possible matches'][index] = all_matches['Possible matches'][index] + append_string + " (found in " + str(int(study_data_combined.sum()[row['Possible matches']])) + " other studies)"
+
+    return all_matches
+
+def attribute_type_prediction(exact_matches, name_matches, value_matches, studies):
+    sample_or_patient = {True:"Patient", False:"Sample"}
+    match_type_table = []
+    only_matches = (pd.concat([exact_matches, value_matches, name_matches])
+                      .drop_duplicates()
+                      .sort_values('New study attribute',axis=0)
+                      .reset_index(drop=True))
+    
+    for matching_attribute in only_matches['Possible matches'].values:
+        studies_with_attribute=study_data_combined[study_data_combined[matching_attribute]>0].index    
```

Comment:
this construction is confusing on first read ... probably because I'm not fluent in the python idioms. I guess the inner selection picks out a set of elements, and the outer selection gives the index numbers of (only) the selected elements.